### PR TITLE
[BE] [FE] Disable opensearch

### PR DESCRIPTION
**Pull request by nialna**
## 🌟 What is the purpose of this PR?

We want to make it easier to run the app locally by making the opensearch docker image optional (because it uses a lot of memory). Search is also currently having issues so we need to disable it anyway.

## ⚠️ Known issues

* Currently opensearch is just commented out in the docker-compose file. I still need to find a better way to add the opensearch service conditionally instead.

## 🐾 Next steps

* We could merge this to disable opensearch (since it's currently broken anyway) and reduce the load on people's computers when running the app. Then we can find a better way of conditionally loading the opensearch container

## 🔍 What does this change?

* A new `HASH_OPENSEARCH_ENABLED` env variable set to true or false allows to control whether search features should be enabled (disabling this will make the `search-loader` exit on start, and will make any GraphQL search request error
* The frontend `search-bar` component is currently commented out since search doesn't work
* The `search` instance in our gql context is now optional
* The `opensearch` image is commented out in our external services `docker-compose` file for now
## 🔗 Related links

- [Way of running app without OpenSearch (disable search if no env for OpenSearch?)](https://app.asana.com/0/1201095311341924/1201472294283856/f)

## 🛡 What tests cover this?

None

## ❓ How to test this?

The search bar won't be present in the frontend. Searching anything would throw an apollo error

### CODE REVIEW

File: external-services/docker-compose.yml [hunk 1]
```diff
@@ -1,31 +1,31 @@
...
+  #     - 9200:9200
+  #   ulimits:
+  #     nofile:
+  #       soft: 65536
+  #       hard: 65536
+  #   ## Mouting open search data to a local directory may lead to java.nio.file.AccessDeniedException.
+  #   ## Details: https://github.com/opensearch-project/OpenSearch/issues/1579.
+  #   ## We can revisit the setup after upgrading base image or by fixing permissions in a custom image.
+  #   # volumes:
+  #   #   - ../var/external-services/opensearch/data:/usr/share/opensearch/data
```
Comment (kachkaev) [inline external-services/docker-compose.yml:28]: Let me try to play around with this a little. Maybe there is a cheap way to create a custom entry point that will enter "noop" mode based on some env var.

**Review by kachkaev — COMMENTED** [review]
Thanks @nialna! Do you know why Playwright tests could be failing?

**Comment by nialna** [comment]
> Thanks @nialna! Do you know why Playwright tests could be failing?

I've been looking into this and it seems to be caused by the verification codes not getting into the clipboard in time when I run it locally... I've been investigating and it happens on main too for me, I was just about to ask if anyone else had this issue

File: packages/hash/playwright/tests/entity-creation.spec.ts [hunk 2]
```diff
@@ -23,8 +23,8 @@ test("user can create and update entity", async ({ page }) => {
   // Fill up entity creation form
   await page.click('text=NameDescription >> input[type="text"]');
   await page.fill('text=NameDescription >> input[type="text"]', entityName);
-  await page.click(':nth-match(input[type="text"], 3)');
-  await page.fill(':nth-match(input[type="text"], 3)', "Test Entity");
+  await page.click(':nth-match(input[type="text"], 2)');
+  await page.fill(':nth-match(input[type="text"], 2)', "Test Entity");
```
Comment (kachkaev) [inline packages/hash/playwright/tests/entity-creation.spec.ts:27]: Hmm interesting and makes sense! If tests pass, how about making this selector less page-aware? We can select some screen area first and then an 'nth' input. Tests will get less fragile.

Comment (nialna) [reply packages/hash/playwright/tests/entity-creation.spec.ts:27]: yeah I was thinking the same. Not sure what's the best way to do this in playwright. Should we give an `id` to the text field in the html so that we can reliably get the input from its id or is there a canonical way of doing this predictably in playwright?

File: packages/hash/playwright/tests/utils/getDerivedPayloadFromMostRecentEmail.ts [hunk 3]
```diff
@@ -31,9 +31,11 @@ const waitForRecentFileChange = async (
     await sleep(checkIntervalInMs);
     remainingWaitTime -= checkIntervalInMs;
   } while (remainingWaitTime > 0);
-
+  const timeLastEdited = (await fs.stat(filePath)).mtimeMs;
```
Comment (kachkaev) [inline packages/hash/playwright/tests/utils/getDerivedPayloadFromMostRecentEmail.ts:34]: This statement may throw if `filePath` does not exist. I guess we can revert this file as it looks unrelated.

File: packages/hash/search-loader/src/index.ts [hunk 4]
```diff
@@ -22,6 +22,13 @@ import { StatsD } from "hot-shots";
 import { getSystemAccountId, getEntityType } from "./db";
 import { logger, INSTANCE_ID } from "./config";
 
+const OPENSEARCH_ENABLED = process.env.HASH_OPENSEARCH_ENABLED === "true";
+if (!OPENSEARCH_ENABLED) {
+  // eslint-disable-next-line no-console
+  console.log("Opensearch is not enabled. Shutting down search-loader");
+  process.exit(1);
```
Comment (kachkaev) [inline packages/hash/search-loader/src/index.ts:29]:
```suggestion
  process.exit(0);
```

`1` means error, but we do as expected here.

File: packages/hash/playwright/tests/entity-creation.spec.ts (hunk 2 above)
Comment (kachkaev) [reply packages/hash/playwright/tests/entity-creation.spec.ts:27]:
```suggestion
  await page.click('main :nth-match(input[type="text"], 2)');
  await page.fill('main :nth-match(input[type="text"], 2)', "Test Entity");
```

Screenshot 2022-01-18 at 18 15 12

This will keep the tests working when we bring back the input. I haven't checked new selectors – my suggestion is just a guess.

Comment (nialna) [reply packages/hash/playwright/tests/entity-creation.spec.ts:27]: oh I see what you mean, that makes more sense. I'll try it this way

File: external-services/opensearch/start.sh [hunk 5]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
+
+if [[ "$HASH_OPENSEARCH_ENABLED" = "true" ]]; then
+  source /usr/share/opensearch/opensearch-docker-entrypoint.sh;
+else
+  echo "Not using Open Search because HASH_OPENSEARCH_ENABLED is not equal to true.";
+  echo "Opensearch is currently disabled by default in Hash to save memory (the opensearch docker image is pretty memory intensive)";
```
Comment (kachkaev) [inline external-services/opensearch/start.sh:7]: We are using `Open Search` on line 1 and `Opensearch` on line 2 (like in `index.ts`):

```
  console.log("Opensearch is not enabled. Shutting down search-loader");
```

Let’s pick one ‘official’ style and stick with it.

Comment (kachkaev) [reply external-services/opensearch/start.sh:7]: There's also OpenSearch and Opensearch in `packages/hash/README.md` 👀

File: packages/hash/frontend/src/components/layout/PageHeader/SearchBar.tsx [hunk 6]
```diff
@@ -97,6 +99,12 @@ export const SearchBar: React.VFC = () => {
   // present loading screen while waiting for the user to stop typing
   const isLoading = loading || displayedQuery !== submittedQuery;
 
+  // /!\ /!\ /!\
+  // NOTE: This component is currently returning an empty div as a way to disable it temporarily due to search being broken
+  // /!\ /!\ /!\
```
Comment (kachkaev) [inline packages/hash/frontend/src/components/layout/PageHeader/SearchBar.tsx:104]:
```suggestion
  // NOTE: This component is currently returning an empty div as a way to disable it temporarily due to search being broken
```

**Comment by nialna** [comment]
Just updated everything and also modified the frontend to use the env variable so we can toggle the whole feature on or off with the same flag

File: packages/hash/frontend/src/components/layout/PageHeader/SearchBar.tsx [hunk 7]
```diff
@@ -97,6 +97,10 @@ export const SearchBar: React.VFC = () => {
   // present loading screen while waiting for the user to stop typing
   const isLoading = loading || displayedQuery !== submittedQuery;
 
+  // NOTE: This component is currently returning an empty div as a way to disable it when opensearch is off
+  if (process.env.NEXT_PUBLIC_HASH_OPENSEARCH_ENABLED !== "true") {
+    return <div />;
+  }
```
Comment (kachkaev) [inline packages/hash/frontend/src/components/layout/PageHeader/SearchBar.tsx:103]: We may end up with a situation some component logic is broken because the server is off. So instead of conditionally returning an empty div, we can make the entire component conditional:

```tsx
const SearchBarWhenSearchIsEnabled = ...;

const SearchBarWhenSearchIsDisabled: VoidFunctionComponent = () => {
  return <div />
}

export const SearchBar = process.env.NEXT_PUBLIC_HASH_OPENSEARCH_ENABLED === "true"
  ? SearchBarWhenSearchIsEnabled
  : SearchBarWhenSearchIsDisabled;
```

We [did something similar for `useCollabPositions()` hook](https://github.com/hashintel/hash/commit/b3aae3e0895f1745e213add9e1c8ee40257daaea#diff-4b88e35f8fb69be9809c05199458282ba9191290039868af762837809acc120aR92-R99).

File: .env [hunk 8]
```diff
@@ -6,6 +6,7 @@ FRONTEND_DOMAIN=localhost:3000
 HASH_COLLAB_QUEUE_NAME=collab
 HASH_EMAIL_TRANSPORTER=dummy
 
+NEXT_PUBLIC_HASH_OPENSEARCH_ENABLED=false
```
Comment (kachkaev) [inline .env:9]: I understand why we use `NEXT_PUBLIC_*` here but I wonder if we can actually avoid it. If we don’t come up with some workaround, then every time we’ll need a new var in UI, we'll have to touch other packages. Seeing `NEXT_PUBLIC_* outside Next.js is a bit confusing.

Since we’ve [managed to ‘feed’ our frontend with the monorepo‘s `.env*` files](https://github.com/hashintel/hash/pull/185), how about doing this in `next.config.js`?

```diff
 const { buildStamp } = require("./buildstamp");

 config({ silent: true, path: "../../.." });
+process.env.NEXT_PUBLIC_X = process.env.X
+process.env.NEXT_PUBLIC_Y = process.env.Y

 const sentryWebpackPluginOptions = {
```

Top-level `.env*` files won’t have to be aware of Next!

**Review by kachkaev — COMMENTED** [review]
👍👍👍

File: .env (hunk 8 above)
Comment (nialna) [reply .env:9]: It is a bit confusing but then again having two different names for env variables would also be. I thought of just making them separate but I don't like the idea of the same having having two different variable names

Comment (kachkaev) [reply .env:9]: We don't have to use `NEXT_PUBLIC_*` vars in multiple places, so we won't have too many cases when a variable is referred under two names. We can add prefixes in `next.config.js` and then remove them, say, in `src/lib/public-env.ts`:

```ts
export const X = process.env.NEXT_PLUBLIC_X;
export const Y = process.env.NEXT_PLUBLIC_Y;
```

That’s not ideal, but probably a bit better than adding `NEXT_PUBLIC_*` outside `packages/hash/frontend`. If we agree to use `NEXT_PUBLIC_*` only in two files (`next.config.js` and `public-env.ts`), maybe it won’t be too bad.

File: external-services/opensearch/start.sh (hunk 5 above)
Comment (nonparibus) [reply external-services/opensearch/start.sh]: OpenSearch is correct.

Comment (nialna) [reply external-services/opensearch/start.sh]: yeah, I've replaced the ones I could find with OpenSearch after checking the website. There might still be older bits of text in other places that still have it wrong though.

File: .env (hunk 8 above)
Comment (nialna) [reply .env]: Doing it this way seems fine then, I've updated it to use this

**Review by kachkaev — APPROVED** [review]
Thanks Liana! 🚀
